### ISSUE TITLE
ntopng 1D view redirects to connection failed

### ISSUE THREAD

**Issue by juesor** [ntop/ntopng#2079]
Package ntopng.x86_64 0:3.7.181002-4897
Package ntopng-data.noarch 0:3.7.181002-4897

When I navigate to Traffic Report and click 1D nothing loads.  

50% of the time i get a page cannot be displayed.

![ntopng 1d unable to load](https://user-images.githubusercontent.com/14931420/46410847-c468a880-c6df-11e8-8919-61907a842804.jpg)

I have updated to the latest nightly and its more of the same the page is slow to load.  

  Verifying  : ntopng-data-3.7.181003-4908.noarch
  Verifying  : ntopng-3.7.181003-4908.x86_64

I can get it to work 50% of the time by clicking 1D and waiting.  Then clicking F5 to refresh and then selecting yesterday and today in the calendar fields and then clicking generate.  

This started after the Oct 01 update but prior to that i was on Sep 19.

Sep 19 01:43:43 Updated: ntopng-data-3.7.180919-4806.noarch
Sep 19 01:43:48 Updated: ntopng-3.7.180919-4806.x86_64
Oct 01 09:05:41 Updated: ntopng-data-3.7.181001-4881.noarch
Oct 01 09:05:46 Updated: ntopng-3.7.181001-4881.x86_64
Oct 02 09:03:47 Updated: ntopng-data-3.7.181002-4897.noarch
Oct 02 09:04:00 Updated: ntopng-3.7.181002-4897.x86_64
Oct 03 08:21:33 Updated: ntopng-data-3.7.181003-4908.noarch
Oct 03 08:21:46 Updated: ntopng-3.7.181003-4908.x86_64

**Comment by simonemainardi** [comment]
Uhm, I am not sure this is due to ntopng.

I can see a connection refused from the browser error. This typically occurs when the host (or someone in the middle like a firewall) sends you a tcp RST. That could be due to a closed port or to a firewall that doesn't want you to do that connection.

With a slow, unresponsive ntopng you would have had the page loading very very slowly or a connection timeout. But the connection refused error is strange...

Even if you open tens of connections to ntopng they will be put in a waiting queue and will eventually be served by the ntopng embedded web-server thread pool, but ntopng per se won't refuse a connection.

Can you please check.

**Comment by juesor** [comment]
I sent you an email with a dropbox link to a video of what is happening.

The ntopng service is crashing and restarting in the process.  So it does not look to be a firewall issue.  This has worked in the past.

● ntopng.service - ntopng high-speed web-based traffic monitoring and analysis tool
   Loaded: loaded (/usr/lib/systemd/system/ntopng.service; disabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: signal) since Wed 2018-10-03 13:05:09 EDT; 3s ago
  Process: 38148 ExecStopPost=/bin/sh -c /bin/echo "$(/bin/date) ntopng StopPost" >> /var/log/ntop-systemd.log (code=exited, status=0/SUCCESS)
  Process: 38146 ExecStopPost=/bin/rm -rf /run/ntopng.conf (code=exited, status=0/SUCCESS)
  Process: 38071 ExecStartPost=/bin/sh -c /bin/echo "$(/bin/date) ntopng StartPost" >> /var/log/ntop-systemd.log (code=exited, status=0/SUCCESS)
  Process: 38070 ExecStart=/usr/local/bin/ntopng /run/ntopng.conf (code=killed, signal=SEGV)
  Process: 38066 ExecStartPre=/bin/sh -c /bin/sed "/^[ ]*-e.*$\|^[ ]*-G.*\|^[ ]*--daemon.*\|^[ ]*--pid.*/s/^/#/" /etc/ntopng/ntopng.conf > /run/ntopng.conf (code=exited,       =0/SUCCESS)
  Process: 38049 ExecStartPre=/bin/sh -c /usr/bin/ntopng-utils-manage-config -a check-restore  && /usr/bin/ntopng-utils-manage-config -a restore || true (code=exited, st       SUCCESS)
  Process: 38043 ExecStartPre=/bin/sh -c /bin/echo "$(/bin/date) ntopng StartPre" >> /var/log/ntop-systemd.log (code=exited, status=0/SUCCESS)
 Main PID: 38070 (code=killed, signal=SEGV)


I showed this in the video it kills the process and restarts automatically.

**Comment by simonemainardi** [comment]
Thanks for providing the video. If the service crashes then a connection refused makes sense.

Please, generate a core dump of the crash and upload it somewhere along with the ntopng binary. Core dump generation instructions can be found at https://www.ntop.org/support/need-help-2/need-help/

Please, also include the ntopng configuration file.

**Comment by simonemainardi** [comment]
also make sure to use version >= 3.7.181004 at this could have affected also this issue https://github.com/ntop/ntopng/commit/4ae68f9e44a94d8c1c0a90d1bca09624fd778a3c

**Comment by juesor** [comment]
Ok you may need to walk me through doing the crash a little more on centos.  But yes i am running on.

3.7.181005-4932

**Comment by simonemainardi** [comment]
OK:

- stop ntopng
- install gdb
- execute "gdb --args /usr/bin/ntopng /etc/ntopng/ntopng.conf"
- type "r" in the gdb console (stands for run)
- use ntopng regularly and wait until it crashes
- go back to the gdb console and save the core file with command "generate-core-file"

**Comment by juesor** [comment]
Ok i am doing that now.

I would suggest you move from mysql to something like elasticsearch i've been working with some graylog2 servers and for this type of application elastic blows mysql out of the water and a lot of my issues with ntopng come from the mysql integration.  I'll post the output of this shortly.

**Comment by simonemainardi** [comment]
Thanks for reporting @juesor. The partial trace you've sent me was helpful, in fact. I have been able to reproduce a crash in the exact point highlighted by your trace.

```
(gdb) bt
#0  0x0000000000482ac5 in AddressTree::match(char*) (this=0x0, addr=0x7fffdb7fba70 "192.168.2.0/24") at src/AddressTree.cpp:132
#1  0x00000000004b09fa in ntop_is_allowed_network(lua_State*) (vm=0x40000378) at src/LuaEngine.cpp:5964
#2  0x00000000005552da in lj_BC_FUNCC ()
#3  0x000000000054475d in lua_pcall ()
#4  0x00000000004ded9b in LuaHandler::luaL_dofileM(bool) (this=0x7fffd0004ee0, execute=true) at pro/LuaHandler.cpp:34
#5  0x00000000004b5b48 in __ntop_lua_handlefile(lua_State*, char*, bool) (L=0x40000378, script_path=0x7fffdb7fc160 "/home/simone/ntopng/scripts/lua/pro/report.lua", ex=true) at src/LuaEngine.cpp:7608
#6  0x00000000004b7b1d in LuaEngine::handle_script_request(mg_connection*, mg_request_info const*, char*, bool*, char const*) (this=0x7fffd0004dd0, conn=0x7fffd00008c0, request_info=0x7fffd00008c0, script_path=0x7fffdb7fc160 "/home/simone/ntopng/scripts/lua/pro/report.lua", attack_attempt=0x7fffdb7fca67, user=0x7fffdb7fc0a0 "nologin") at src/LuaEngine.cpp:8716
#7  0x000000000042ae74 in handle_lua_request(mg_connection*) (conn=0x7fffd00008c0) at src/HTTPserver.cpp:889
#8  0x00000000004251df in handle_request(mg_connection*) (conn=0x7fffd00008c0) at src/../third-party/mongoose/mongoose.c:4412
#9  0x000000000042734b in process_new_connection(mg_connection*) (conn=0x7fffd00008c0) at src/../third-party/mongoose/mongoose.c:5120
#10 0x000000000042788e in worker_thread(void*) (thread_func_param=0x4dd9af0) at src/../third-party/mongoose/mongoose.c:5214
#11 0x00007ffff6b4de25 in start_thread () at /lib64/libpthread.so.0
#12 0x00007ffff4c5ebad in clone () at /lib64/libc.so.6
```

I've pushed a fix (https://github.com/ntop/ntopng/commit/0cf2d567865669b0bae89268988e894d8fcc0a46). A new build is in progress and will be available in 1h from now. Please, let me know. If you still experience the crash, then please try and upload the whole ntopng core (see instructions above) and not just the stacktrace with the latest call.

Thank you.

**Comment by juesor** [comment]
I ran the update this morning and am currently on 

Updated:
  ntopng.x86_64 0:3.7.181008-4953            ntopng-data.noarch 0:3.7.181008-4953            pfring.x86_64 0:7.3.0-2168            pfring-dkms.noarch 0:7.3.0-2168
Updated:
  n2disk.x86_64 0:3.1.181008-5001    ntopng.x86_64 0:3.7.181008-4953    ntopng-data.noarch 0:3.7.181008-4953    pfring.x86_64 0:7.3.0-2168    pfring-dkms.noarch 0:7.3.0-2168


But every time update one of my servers will sit here for every time it sits at this screen for hours.

Yes my DB is.

Filesystem           Size  Used Avail Use% Mounted on
/dev/sdd1            394G  194G  181G  52% /var/lib/mysql

194G 


Database schema migration in progress, waiting for database ntopng to become operational. You will be redirected as soon as the database is ready. Depending on the size of the existing database, this operation can take a long time. Migrations are performed only once and only during major version upgrades. If in doubt, please contact the developers at info@ntop.org.

**Comment by juesor** [comment]
2 hours later still waiting on 1 server but first server is working.

**Comment by simonemainardi** [comment]
Thanks for reporting. I'm glad to read the issue with the reports is solved. It is safe to close this issue now. Thank you.

As per the database schema migration, this occurs when the mysql table schema needs some changes. This is a sporadic update operation and we haven't been touching the schema for more than 1 year now. So I would expect it not to happen at every update. If you are getting this after every update, please, file another issue and report all the details you have.
